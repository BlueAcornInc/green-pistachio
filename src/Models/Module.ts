type ModuleConstructorArgs = {
    sourceDirectory: string;
    name: string;
    enabled?: boolean;
};

export default class Module {
    private name: string;
    private sourceDirectory: string;
    private enabled: boolean;

    constructor(config: ModuleConstructorArgs) {
        const { name, sourceDirectory, enabled } = config;

        this.name = name;
        this.sourceDirectory = sourceDirectory;
        this.enabled = enabled || false;
    }

    public getName() {
        return this.name;
    }

    public getSourceDirectory() {
        return this.sourceDirectory;
    }

    public getEnabled() {
        return this.enabled;
    }

    public setEnabled(enabled: boolean) {
        this.enabled = enabled;
    }
}