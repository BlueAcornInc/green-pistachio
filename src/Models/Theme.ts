import glob from 'fast-glob';
import debug from 'debug';
const logger = debug('gpc:theme');

type ThemeConstructorArgs = ThemeData & {
    sourceDirectory: string;
    parent?: Theme;
};

// type CriticalCssInput = {
//     filepath: string;
//     urls: string[];
// };

export type ThemeData = {
    area: "frontend" | "adminhtml";
    path: string;
    enabled?: boolean;
    locales?: string[];
    stylesheets?: string[];
    // criticalCss?: CriticalCssInput[];
};

export default class Theme {
    private sourceDirectory: string;
    private parent?: Theme;
    private enabled: boolean | undefined;
    private data: ThemeData;

    constructor(config: ThemeConstructorArgs) {
        const {
            sourceDirectory,
            area,
            path,
            parent,
            locales,
            stylesheets,
            // criticalCss,
            enabled
        } = config;

        this.sourceDirectory = sourceDirectory;
        this.parent = parent;
        this.enabled = enabled || false;
        this.data = {
            area,
            path,
            locales,
            stylesheets,
            // criticalCss
        };
    }

    public configure(data: Partial<ThemeData>) {
        this.data = {
            ...this.data,
            ...data
        };
    }

    public getSourceDirectory() {
        return this.sourceDirectory;
    }

    public getParent() {
        return this.parent;
    }

    public getData() {
        return this.data;
    }

    public getLocales(): string[] {
        if (this.data.locales) {
            return this.data.locales;
        }

        logger(`${this.data.path}: Can't Determine Locales, defaulting to en_US`);

        return ['en_US'];
    }

    public getStyleSheets(): string[] {
        if (this.data.stylesheets) {
            return this.data.stylesheets;
        }

        logger(`${this.data.path}: Stylesheets not explicitly defined, detecting entries now.`);

        const files: Set<string> = new Set();

        const discoveredFiles = glob
            .sync(`${this.getSourceDirectory()}/**/[^_]*.less`)
            .map(sourceFile => sourceFile.replace(`${this.getSourceDirectory()}/web/`, ''));

        for (const file of discoveredFiles) {
            files.add(file);
        }

        let childTheme = this.getParent();

        while (childTheme) {
            for (const file of childTheme.getStyleSheets()) {
                files.add(file);
            }
            childTheme = childTheme.getParent();
        }

        const stylesheets = Array.from(files);

        logger(`${this.data.path}: Discovered the following stylesheets: ${stylesheets.join(', ')}`);

        return stylesheets;
    }

    // public getCriticalPaths(): CriticalCssInput[] {
    //     if (this.data.criticalCss) {
    //         return this.data.criticalCss;
    //     }

    //     logger(`No critical path data available for: ${this.data.path}`);

    //     return [];
    // }

    public getEnabled() {
        return this.enabled;
    }

    public setEnabled(enabled: undefined | boolean) {
        this.enabled = enabled;
    }
}
