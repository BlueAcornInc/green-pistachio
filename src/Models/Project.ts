import { join } from 'path';
import { SyncHook } from 'tapable';
import debug from 'debug';
import Module from "./Module";
import Theme, { ThemeData } from "./Theme";
import ConfigLoader from './Project/ConfigLoader';
import { Configuration } from 'webpack';
import WebpackConfigFactory from '../Webpack/ConfigFactory';
const logger = debug('gpc:project');

type ProjectConstructorArgs = {
    root: string;
    themes: Theme[];
    modules: Module[];
    enabledThemes?: string[];
};

type WebpackDevServerConfig = {
    proxyUrl: string;
    devServerUrl: string;
    devServerPort: number;
};

export default class Project {
    public root: string;
    private themes: Theme[];
    private enabledThemes: Theme[];
    private modules: Module[];
    private webpackDevelopmentMode: boolean = false;
    private proxyUrl: string = '';
    private devServerConfig: Partial<WebpackDevServerConfig> = {};

    public hooks = {
        // These hooks allow for modifying various gulp configurations
        gulp: {
            babelConfig: new SyncHook(["config"]),
            babelTypeScriptConfig: new SyncHook(["config"]),
            pngSpriteConfig: new SyncHook(["config"]),
            svgSpriteConfig: new SyncHook(["config"]),
            eslintConfig: new SyncHook(["config"]),
            criticalCssConfig: new SyncHook(["config"]),
        },
        webpack: {
            config: new SyncHook(["config"]),
            entryGlobs: new SyncHook(["globs"]),
            appendConfig: new SyncHook<[Configuration[], WebpackConfigFactory]>(["webpackConfigs", "configFactory"])
        },
        configure: new SyncHook(["project"])
    };

    public experiments = {
        webpack: {
            // Set to true in the the project.configure hook to enable unstable HMR functionality
            hmr: false,
            // Set to true to enable css modules
            cssModules: false,
            // Set to true to emit files to the pub directory instead of the individual theme directories
            emitFilesToPub: false,
        }
    };

    constructor(config: ProjectConstructorArgs) {
        const { root, themes, modules, enabledThemes } = config;

        this.root = root;
        this.themes = themes;
        this.modules = modules;
        this.enabledThemes = enabledThemes
            ? themes.filter(theme => enabledThemes.includes(theme.getData().path))
            : [];
    }

    public configure() {
        if (this.enabledThemes.length > 0) {
            // Enable default themes
            for (const theme of this.enabledThemes) {
                theme.setEnabled(true);
            }
        } else {
            for (const theme of this.themes) {
                if (
                    !theme
                        .getSourceDirectory()
                        .includes(
                            join(
                                this.getRootDirectory(),
                                'vendor'
                            )
                        )
                ) {
                    theme.setEnabled(true);
                    this.enabledThemes.push(theme);
                }
            }
        }

        // Enable default modules
        for (const magentoModule of this.modules) {
            if (
                !magentoModule
                    .getSourceDirectory()
                    .includes(
                        join(
                            this.getRootDirectory(),
                            'vendor'
                        )
                    )
            ) {
                magentoModule.setEnabled(true);
            }
        }

        this.hooks.configure.call(this);
    }

    public configureTheme(themeData: Partial<ThemeData>) {
        if (!themeData.path) {
            logger(`can not configure theme without defined path: ${JSON.stringify(themeData)}`);
            return;
        }

        const theme = this.themes.find(theme => theme.getData().path === themeData.path);

        if (!theme) {
            logger(`can not find theme with path = '${themeData.path}'`);
            return;
        }

        theme.configure(themeData);

        if (themeData.hasOwnProperty('enabled')) {
            theme.setEnabled(themeData.enabled);
        }
    }

    public enableModule(moduleName: string) {
        const magentoModule = this.modules.find(magentoModule => magentoModule.getName() === moduleName);

        if (!magentoModule) {
            logger(`could not find module by name: ${moduleName}`);
            return;
        }

        magentoModule.setEnabled(true);
    }

    public getThemes() {
        return this.enabledThemes.filter(magentoTheme => magentoTheme.getEnabled());
    }

    public getAllThemes() {
        return this.themes;
    }

    public getModules() {
        return this.modules.filter(magentoModule => magentoModule.getEnabled());
    }

    public getAllModules() {
        return this.modules;
    }

    public getRootDirectory() {
        return this.root;
    }

    public getThemePubDirectories(theme: Theme) {
        return theme.getLocales().map(locale => join(
            this.getStaticDirectory(),
            theme.getData().area,
            theme.getData().path,
            locale
        ));
    }

    public getWebpackOutputDirectories(theme: Theme) {
        if (this.isWebpackDevelopmentMode() || this.experiments.webpack.emitFilesToPub) {
            return theme.getLocales().map(locale => join(
                this.getRootDirectory(),
                ...(
                    this.isWebpackDevelopmentMode()
                    ? ['static']
                    : ['pub', 'static']
                ),
                theme.getData().area,
                theme.getData().path,
                locale
            ));
        }

        return [join(theme.getSourceDirectory(), 'web')];
    }

    public getStaticDirectory() {
        return join(
            this.getRootDirectory(),
            'pub',
            'static'
        );
    }

    public isWebpackDevelopmentMode() {
        return this.webpackDevelopmentMode;
    }

    public setWebpackDevelopmentMode() {
        this.webpackDevelopmentMode = true;
    }

    public setProxyUrl(url: string) {
        this.proxyUrl = url;
    }

    public getProxyUrl(): string {
        if (this.proxyUrl.length === 0) {
            if (!debug.enabled('gpc:logger')) {
                debug.enable('gpc:logger');
            }
            logger(`A project URL must be provided in your projects ${ConfigLoader.CONFIG_FILE}`);
            logger(`\n\nExample:
module.exports = project => {
project.hooks.configure.tap("Set Product Url", () => {
    project.setProxyUrl("https://example.test");
});
}
            `);

            throw new Error(`A project URL must be provided in your projects ${ConfigLoader.CONFIG_FILE}`);
        }

        return this.proxyUrl;
    }

    public setWebpackDevServerConfig(config: Partial<WebpackDevServerConfig>) {
        if (config.proxyUrl) {
            this.proxyUrl = config.proxyUrl;
        }

        this.devServerConfig = config;
    }

    public getWebpackDevServerConfig(): WebpackDevServerConfig {
        return {
            proxyUrl: this.getProxyUrl(),
            devServerUrl: 'green-pistachio.test',
            devServerPort: 8080,
            ...(this.devServerConfig || {})
        };
    }
}
