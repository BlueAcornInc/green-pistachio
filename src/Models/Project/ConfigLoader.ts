import { promises as fs } from 'fs';
import { join } from 'path';
import debug from 'debug';
import Project from "../Project";
const logger = debug('gpc:configLoader');

export default class ConfigLoader {
    public static CONFIG_FILE = 'green-pistachio.config.js';

    constructor(private project: Project) {}

    get filePath(): string {
        return join(
            this.project.getRootDirectory(),
            ConfigLoader.CONFIG_FILE
        );
    }

    public async configExists(): Promise<boolean> {
        try {
            const configExists = await fs.stat(this.filePath);

            if (configExists) {
                return true;
            }
        } catch (err) {
            logger(`Config file can't be found at ${this.filePath}`);
        }

        return false;
    }

    public async loadConfig(): Promise<boolean> {
        const configExists = await this.configExists();

        if (configExists) {
            try {
                this.interopRequire()(this.project);
                return true;
            } catch (err) {
                logger(`Problem occurred while running user provided configuration file: ${err}`);
            }
        }

        return false;
    }

    private interopRequire() {
        const configModule = require(this.filePath);
        return configModule && configModule.__esModule ? configModule.default : configModule;
    }
}