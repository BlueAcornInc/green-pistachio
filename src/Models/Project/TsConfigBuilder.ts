import { promises as fs } from 'fs';
import { join } from 'path';
import debug from 'debug';
import Project from "../Project";
const logger = debug('gpc:typescript:configBuilder');

export default class TsConfigBuilder {
    public getFilePath(): string {
        return join(__dirname, '..', '..', '..', 'green-pistachio.tsconfig.json');
    }

    public async configFileExists(): Promise<boolean> {
        try {
            await fs.stat(this.getFilePath());

            return true;
        } catch (err) {
            return false;
        }
    }

    public async emitConfigFile(project: Project) {
        logger(`Generating Projects tsconfig.json base`);
        const include: string[] = [];
        const paths: Record<string, string[]> = {}

        for (const module of project.getModules()) {
            const moduleSrcDir = module.getSourceDirectory().replace(project.getRootDirectory(), '');

            const modulePaths: string[] = [
                join(moduleSrcDir, 'view/frontend/web/ts/*'),
                join(moduleSrcDir, 'view/base/web/ts/*'),
                join(moduleSrcDir, 'view/adminhtml/web/ts/*'),
            ];

            paths[`${module.getName()}/*`] = modulePaths.map(modulePath => modulePath.replace(/^\//, ''));

            include.push(join(moduleSrcDir, 'view/frontend/web/ts/**/*'));
            include.push(join(moduleSrcDir, 'view/base/web/ts/**/*'));
            include.push(join(moduleSrcDir, 'view/adminhtml/web/ts/**/*'));
        }

        for (const theme of project.getAllThemes()) {
            const themeSrcDir = theme.getSourceDirectory().replace(project.getRootDirectory(), '');

            include.push(join(themeSrcDir, '**/*'));
            
            // Add theme overrides to module paths
            for (const moduleName of Object.keys(paths)) {
                paths[moduleName].unshift(
                    join(themeSrcDir, moduleName.replace(/\/\*$/, ''), 'web/ts/*').replace(/^\//, '')
                );
            }
        }

        logger(`Resolved include paths to: ${JSON.stringify(include)}`);
        logger(`Resolved module paths to: ${JSON.stringify(paths)}`);

        const writePath = this.getFilePath();
        const config = this.getBaseConfig();

        config.include = include.map(includePath => `./${includePath}`);
        config.compilerOptions.paths = paths;

        try {
            logger(`Writing config file: ${writePath}`);
            await fs.writeFile(writePath, JSON.stringify(config, null, 4));
        } catch (err) {
            logger(`Problem writing config file: ${writePath}`);
            throw err;
        }

        return writePath;
    }

    private getBaseConfig() {
        return {
            "include": [
              "./app/**/*"
            ],
            "compilerOptions": {
              "baseUrl": ".",
              "paths": {},
              "strict": true,
              "esModuleInterop": true,
              "lib": [
                "dom",
                "es2015"
              ],
              "jsx": "react",
              "emitDecoratorMetadata": true,
              "experimentalDecorators": true,
              "target": "es5",
              "allowJs": true,
              "skipLibCheck": true,
              "allowSyntheticDefaultImports": true,
              "forceConsistentCasingInFileNames": true,
              "module": "esnext",
              "moduleResolution": "node",
              "resolveJsonModule": true,
              "isolatedModules": true,
              "noEmit": true
            }
          }
    }
}