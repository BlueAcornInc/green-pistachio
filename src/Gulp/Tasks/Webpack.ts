import { TaskFunction, watch } from "gulp";
import debug from 'debug';
import WebpackDevServer from 'webpack-dev-server';
import { certificateFor } from 'devcert';
import Project from "../../Models/Project";
import Theme from "../../Models/Theme";
import WebpackConfigFactory from "../../Webpack/ConfigFactory";
import GetCompiler from "../../Webpack/GetCompiler";
import { TaskInterface } from "./TaskInterface";
import TsConfigBuilder from "../../Models/Project/TsConfigBuilder";
import { responseInterceptor } from "../../Webpack/DevServerResponseInterceptor";
const logger = debug('gpc:gulp:webpack');

export default class Webpack implements TaskInterface {
    private webpackConfigFactory: WebpackConfigFactory;
    private getCompiler: GetCompiler;
    private tsConfigBuilder: TsConfigBuilder;
    private configFileExists: boolean = false;

    constructor() {
        this.webpackConfigFactory = new WebpackConfigFactory();
        this.getCompiler = new GetCompiler();
        this.tsConfigBuilder = new TsConfigBuilder();
    }

    execute(project: Project, theme?: Theme): TaskFunction {
        const webpackTask: TaskFunction = async (done) => {
            if (!this.configFileExists) {
                const exists = await this.tsConfigBuilder.configFileExists();
                
                if (!exists) {
                    await this.tsConfigBuilder.emitConfigFile(project);
                }

                this.configFileExists = true;
            }

            const config = await this.webpackConfigFactory.getConfig(project, theme);
            const compiler = this.getCompiler.execute(config);

            compiler.run((err, stats) => {
                if (err) {
                    logger(err);
                }

                if (stats) {
                    logger(stats.toString({
                        colors: true
                    }));
                }

                done();
            });
        };

        webpackTask.displayName = 'webpack';

        return webpackTask;
    }

    watch(project: Project, theme?: Theme): TaskFunction {
        return async (done) => {
            project.setWebpackDevelopmentMode();
            const config = await this.webpackConfigFactory.getConfig(project, theme);
            const compiler = this.getCompiler.execute(config);

            const {
                proxyUrl,
                devServerUrl,
                devServerPort
            } = project.getWebpackDevServerConfig();

            let frontendDevServerUrl = `https://${devServerUrl}`;

            if (devServerPort !== 80) {
                frontendDevServerUrl += `:${devServerPort}`
            }

            let ssl = await certificateFor(devServerUrl);

            // Store magento cookies and append back with each proxy request
            let cookieData: string[] | null;

            // @ts-ignore
            const server = new WebpackDevServer(compiler, {
                https: {
                    key: ssl.key,
                    cert: ssl.cert
                },
                ...(project.experiments.webpack.hmr ? {
                    hot: true,
                } : {}),
                proxy: {
                    '/': {
                        target: proxyUrl,
                        secure: false,
                        changeOrigin: true,
                        autoRewrite: true,
                        selfHandleResponse: true,
                        onProxyReq: (proxyReq) => {
                            if (cookieData) {
                                proxyReq.setHeader('cookie', cookieData);
                            }
                        },
                        onProxyRes: responseInterceptor(async (buffer, proxyRes, req, res) => {
                            const proxyCookie = proxyRes.headers['set-cookie'];
                            if (proxyCookie) {
                                cookieData = proxyCookie;
                            }

                            if (proxyRes.headers['location']) {
                                res.setHeader(
                                    'location',
                                    proxyRes.headers['location'].replace(proxyUrl, frontendDevServerUrl)
                                );
                            }

                            if ((proxyRes.headers['content-type'] || '').includes('text/html')) {
                                let response =  buffer.toString()
                                    .replace(new RegExp(proxyUrl, 'g'), frontendDevServerUrl)
                                    .replace(
                                        new RegExp(
                                            proxyUrl.replace(/[^a-z0-9,\._]/iug, (a) => 
                                                `\\\\u00${a.charCodeAt(0).toString(16).toUpperCase()}`
                                            ),
                                            'g'
                                        ),
                                        frontendDevServerUrl
                                    )
                                    .replace(
                                        new RegExp(
                                            proxyUrl.replace(/\//g, '\\\\/'),
                                            'g'
                                        ),
                                        frontendDevServerUrl
                                    )

                                return response;
                            }

                            return buffer;
                        })
                    }
                }
            });

            server.listen(devServerPort, devServerUrl, () => {
                logger('dev server started');
            });

            done();
        };
    }
}