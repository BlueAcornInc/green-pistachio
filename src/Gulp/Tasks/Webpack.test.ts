/**
 * @jest-environment node
 */
import { join } from "path";
import { promises as fs } from "fs";
import Project from "../../Models/Project";
import Theme from "../../Models/Theme";
import Webpack from "./Webpack";
import Module from "../../Models/Module";
import mock from "mock-fs";

afterEach(() => {
    mock.restore();
});

const root = join(__dirname, "__fixtures__", "webpack");

const prepareProject = (themes: Theme[]) => {
    return new Project({
        themes,
        modules: [
            new Module({
                name: "BlueAcorn_Module",
                sourceDirectory: `${root}/app/code/BlueAcorn/Module`,
                enabled: true,
            }),
        ],
        root,
    });
};

describe("Gulp: Webpack", () => {
    it("should compile webpack", () => {});

    // TODO: Rewrite Webpack Tests
    it('should compile webpack', async () => {
        const expectedOutputFiles = [
            `${root}/app/code/BlueAcorn/Module/view/frontend/web/js/module.bundle.js`,
            `${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/test-ts.bundle.js`,
            `${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/test.bundle.js`,
            `${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/BlueAcorn_Module/theme.bundle.js`,
            `${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/BlueAcorn_Module/module.bundle.js`,
            `${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/dynamic-import.js`,
            `${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/commons.js`,
            `${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/manifest.json`,
            `${root}/app/design/frontend/BlueAcorn/site/requirejs-config.js`,
        ];

        const themes = [
            new Theme({
                sourceDirectory: `${root}/app/design/frontend/BlueAcorn/site`,
                area: 'frontend',
                path: 'BlueAcorn/site',
                enabled: true
            })
        ];

        const project = prepareProject(themes);

        const webpack = new Webpack();
        webpack.execute(project)(async () => {
            for (const expectedOutputFile of expectedOutputFiles) {
                const contents = await fs.readFile(expectedOutputFile);
                expect(contents.toString()).toMatchSnapshot();
                await fs.unlink(expectedOutputFile)
            }
        });
    });

    it('should update existing requirejs-config.js file #1', async () => {
        const themes = [
            new Theme({
                sourceDirectory: `${root}/app/design/frontend/BlueAcorn/siteExistingRjs`,
                area: 'frontend',
                path: 'BlueAcorn/site',
                enabled: true
            })
        ];

        const project = prepareProject(themes);
        const rjsfile = `${root}/app/design/frontend/BlueAcorn/siteExistingRjs/requirejs-config.js`;
        const originalContent = await fs.readFile(rjsfile);

        const webpack = new Webpack();
        webpack.execute(project)(async () => {
            const contents = await fs.readFile(rjsfile);
            expect(contents.toString()).toMatchSnapshot();
            await fs.writeFile(rjsfile, originalContent.toString());
            await fs.unlink(`${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/BlueAcorn_Module/module.bundle.js`);
            await fs.unlink(`${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/manifest.json`);
        });
    });

    it('should update existing requirejs-config.js file #2', async () => {
        const themes = [
            new Theme({
                sourceDirectory: `${root}/app/design/frontend/BlueAcorn/siteExistingRjsWithManifest`,
                area: 'frontend',
                path: 'BlueAcorn/site',
                enabled: true
            })
        ];

        const project = prepareProject(themes);
        const rjsfile = `${root}/app/design/frontend/BlueAcorn/siteExistingRjsWithManifest/requirejs-config.js`;
        const originalContent = await fs.readFile(rjsfile);

        const webpack = new Webpack();
        webpack.execute(project)(async () => {
            const contents = await fs.readFile(rjsfile);
            expect(contents.toString()).toMatchSnapshot();
            await fs.writeFile(rjsfile, originalContent.toString());
            await fs.unlink(`${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/BlueAcorn_Module/module.bundle.js`);
            await fs.unlink(`${root}/pub/static/frontend/BlueAcorn/site/en_US/bundle/manifest.json`);
        });
    });
});
