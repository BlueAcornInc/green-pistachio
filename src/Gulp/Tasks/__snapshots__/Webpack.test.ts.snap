// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gulp: Webpack should compile webpack 1`] = `"define((()=>(()=>{\\"use strict\\";return{}})()));"`;

exports[`Gulp: Webpack should compile webpack 2`] = `"define([\\"./commons\\"],(()=>(()=>{\\"use strict\\";var r,e={649:(r,e,t)=>{t(979)}},t={};function n(r){var o=t[r];if(void 0!==o)return o.exports;var i=t[r]={exports:{}};return e[r](i,i.exports,n),i.exports}n.m=e,r=[],n.O=(e,t,o,i)=>{if(!t){var a=1/0;for(f=0;f<r.length;f++){for(var[t,o,i]=r[f],u=!0,l=0;l<t.length;l++)(!1&i||a>=i)&&Object.keys(n.O).every((r=>n.O[r](t[l])))?t.splice(l--,1):(u=!1,i<a&&(a=i));u&&(r.splice(f--,1),e=o())}return e}i=i||0;for(var f=r.length;f>0&&r[f-1][2]>i;f--)r[f]=r[f-1];r[f]=[t,o,i]},n.n=r=>{var e=r&&r.__esModule?()=>r.default:()=>r;return n.d(e,{a:e}),e},n.d=(r,e)=>{for(var t in e)n.o(e,t)&&!n.o(r,t)&&Object.defineProperty(r,t,{enumerable:!0,get:e[t]})},n.g=function(){if(\\"object\\"==typeof globalThis)return globalThis;try{return this||new Function(\\"return this\\")()}catch(r){if(\\"object\\"==typeof window)return window}}(),n.o=(r,e)=>Object.prototype.hasOwnProperty.call(r,e),n.p=\\"\\",(()=>{var r={204:0};n.O.j=e=>0===r[e];var e=(e,t)=>{var o,i,[a,u,l]=t,f=0;for(o in u)n.o(u,o)&&(n.m[o]=u[o]);for(l&&l(n),e&&e(t);f<a.length;f++)i=a[f],n.o(r,i)&&r[i]&&r[i][0](),r[a[f]]=0;n.O()},t=self.webpackChunk=self.webpackChunk||[];t.forEach(e.bind(null,0)),t.push=e.bind(null,t.push.bind(t))})();var o=n.O(void 0,[351],(()=>n(649)));return n.O(o)})()));"`;

exports[`Gulp: Webpack should compile webpack 3`] = `"define([\\"./commons\\"],(()=>(()=>{\\"use strict\\";var e,r,t={971:(e,r,t)=>{t(979),console.log(\\"hello world\\")}},o={};function n(e){var r=o[e];if(void 0!==r)return r.exports;var i=o[e]={exports:{}};return t[e](i,i.exports,n),i.exports}n.m=t,e=[],n.O=(r,t,o,i)=>{if(!t){var a=1/0;for(d=0;d<e.length;d++){for(var[t,o,i]=e[d],l=!0,u=0;u<t.length;u++)(!1&i||a>=i)&&Object.keys(n.O).every((e=>n.O[e](t[u])))?t.splice(u--,1):(l=!1,i<a&&(a=i));l&&(e.splice(d--,1),r=o())}return r}i=i||0;for(var d=e.length;d>0&&e[d-1][2]>i;d--)e[d]=e[d-1];e[d]=[t,o,i]},n.n=e=>{var r=e&&e.__esModule?()=>e.default:()=>e;return n.d(r,{a:r}),r},n.d=(e,r)=>{for(var t in r)n.o(r,t)&&!n.o(e,t)&&Object.defineProperty(e,t,{enumerable:!0,get:r[t]})},n.f={},n.e=e=>Promise.all(Object.keys(n.f).reduce(((r,t)=>(n.f[t](e,r),r)),[])),n.u=e=>\\"dynamic-import.js\\",n.g=function(){if(\\"object\\"==typeof globalThis)return globalThis;try{return this||new Function(\\"return this\\")()}catch(e){if(\\"object\\"==typeof window)return window}}(),n.o=(e,r)=>Object.prototype.hasOwnProperty.call(e,r),r={},n.l=(e,t,o,i)=>{if(r[e])r[e].push(t);else{var a,l;if(void 0!==o)for(var u=document.getElementsByTagName(\\"script\\"),d=0;d<u.length;d++){var s=u[d];if(s.getAttribute(\\"src\\")==e){a=s;break}}a||(l=!0,(a=document.createElement(\\"script\\")).charset=\\"utf-8\\",a.timeout=120,n.nc&&a.setAttribute(\\"nonce\\",n.nc),a.src=e),r[e]=[t];var c=(t,o)=>{a.onerror=a.onload=null,clearTimeout(f);var n=r[e];if(delete r[e],a.parentNode&&a.parentNode.removeChild(a),n&&n.forEach((e=>e(o))),t)return t(o)},f=setTimeout(c.bind(null,void 0,{type:\\"timeout\\",target:a}),12e4);a.onerror=c.bind(null,a.onerror),a.onload=c.bind(null,a.onload),l&&document.head.appendChild(a)}},n.r=e=>{\\"undefined\\"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:\\"Module\\"}),Object.defineProperty(e,\\"__esModule\\",{value:!0})},n.p=\\"\\",(()=>{var e={87:0};n.f.j=(r,t)=>{var o=n.o(e,r)?e[r]:void 0;if(0!==o)if(o)t.push(o[2]);else{var i=new Promise(((t,n)=>o=e[r]=[t,n]));t.push(o[2]=i);var a=n.p+n.u(r),l=new Error;n.l(a,(t=>{if(n.o(e,r)&&(0!==(o=e[r])&&(e[r]=void 0),o)){var i=t&&(\\"load\\"===t.type?\\"missing\\":t.type),a=t&&t.target&&t.target.src;l.message=\\"Loading chunk \\"+r+\\" failed.\\\\n(\\"+i+\\": \\"+a+\\")\\",l.name=\\"ChunkLoadError\\",l.type=i,l.request=a,o[1](l)}}),\\"chunk-\\"+r,r)}},n.O.j=r=>0===e[r];var r=(r,t)=>{var o,i,[a,l,u]=t,d=0;for(o in l)n.o(l,o)&&(n.m[o]=l[o]);for(u&&u(n),r&&r(t);d<a.length;d++)i=a[d],n.o(e,i)&&e[i]&&e[i][0](),e[a[d]]=0;n.O()},t=self.webpackChunk=self.webpackChunk||[];t.forEach(r.bind(null,0)),t.push=r.bind(null,t.push.bind(t))})();var i=n.O(void 0,[351],(()=>n(971)));return n.O(i)})()));"`;

exports[`Gulp: Webpack should compile webpack 4`] = `"define([\\"../commons\\"],(()=>(()=>{\\"use strict\\";var r,e={148:(r,e,t)=>{t(979)}},t={};function n(r){var o=t[r];if(void 0!==o)return o.exports;var i=t[r]={exports:{}};return e[r](i,i.exports,n),i.exports}n.m=e,r=[],n.O=(e,t,o,i)=>{if(!t){var a=1/0;for(f=0;f<r.length;f++){for(var[t,o,i]=r[f],u=!0,l=0;l<t.length;l++)(!1&i||a>=i)&&Object.keys(n.O).every((r=>n.O[r](t[l])))?t.splice(l--,1):(u=!1,i<a&&(a=i));u&&(r.splice(f--,1),e=o())}return e}i=i||0;for(var f=r.length;f>0&&r[f-1][2]>i;f--)r[f]=r[f-1];r[f]=[t,o,i]},n.n=r=>{var e=r&&r.__esModule?()=>r.default:()=>r;return n.d(e,{a:e}),e},n.d=(r,e)=>{for(var t in e)n.o(e,t)&&!n.o(r,t)&&Object.defineProperty(r,t,{enumerable:!0,get:e[t]})},n.g=function(){if(\\"object\\"==typeof globalThis)return globalThis;try{return this||new Function(\\"return this\\")()}catch(r){if(\\"object\\"==typeof window)return window}}(),n.o=(r,e)=>Object.prototype.hasOwnProperty.call(r,e),n.p=\\"\\",(()=>{var r={146:0};n.O.j=e=>0===r[e];var e=(e,t)=>{var o,i,[a,u,l]=t,f=0;for(o in u)n.o(u,o)&&(n.m[o]=u[o]);for(l&&l(n),e&&e(t);f<a.length;f++)i=a[f],n.o(r,i)&&r[i]&&r[i][0](),r[a[f]]=0;n.O()},t=self.webpackChunk=self.webpackChunk||[];t.forEach(e.bind(null,0)),t.push=e.bind(null,t.push.bind(t))})();var o=n.O(void 0,[351],(()=>n(148)));return n.O(o)})()));"`;

exports[`Gulp: Webpack should compile webpack 5`] = `"define((()=>(()=>{\\"use strict\\";return{}})()));"`;

exports[`Gulp: Webpack should compile webpack 6`] = `"(self.webpackChunk=self.webpackChunk||[]).push([[334],{872:(s,t,e)=>{\\"use strict\\";e.r(t),e.d(t,{default:()=>i});const i=\\"This is a dynamic import, it should output into it's own file. See: test.bundle.js\\"}}]);"`;

exports[`Gulp: Webpack should compile webpack 7`] = `"define((()=>(self.webpackChunk=self.webpackChunk||[]).push([[351],{979:(e,s,n)=>{n.p=\`\${n.g.requirejs.s.contexts._.config.baseUrl}/bundle/\`}}])));"`;

exports[`Gulp: Webpack should compile webpack 8`] = `
"{
  \\"test-ts.bundle.js\\": \\"test-ts.bundle.js\\",
  \\"test.bundle.js\\": \\"test.bundle.js\\",
  \\"BlueAcorn_Module/theme.bundle.js\\": \\"BlueAcorn_Module/theme.bundle.js\\",
  \\"BlueAcorn_Module/module.bundle.js\\": \\"BlueAcorn_Module/module.bundle.js\\",
  \\"dynamic-import.js\\": \\"dynamic-import.js\\",
  \\"commons.js\\": \\"commons.js\\"
}"
`;

exports[`Gulp: Webpack should compile webpack 9`] = `
"var config = {};

/** START: MagentoRequireJsManifestPlugin */
config = config || {};
config.map = (config.map || {});
config.map['*'] = (config.map['*'] || {});
config.map['*']['BlueAcorn_Module/js/module.bundle'] = 'bundle/BlueAcorn_Module/module.bundle';
/** END: MagentoRequireJsManifestPlugin */"
`;

exports[`Gulp: Webpack should update existing requirejs-config.js file #1 1`] = `
"var config = {
    deps: [
        \\"js/some-dep\\"
    ]
};

/** START: MagentoRequireJsManifestPlugin */
config = config || {};
config.map = (config.map || {});
config.map['*'] = (config.map['*'] || {});
config.map['*']['BlueAcorn_Module/js/module.bundle'] = 'bundle/BlueAcorn_Module/module.bundle';
/** END: MagentoRequireJsManifestPlugin */"
`;

exports[`Gulp: Webpack should update existing requirejs-config.js file #2 1`] = `
"var config = {
    deps: [
        'js/some-file'
    ]
};

/** START: MagentoRequireJsManifestPlugin */
config = config || {};
config.map = (config.map || {});
config.map['*'] = (config.map['*'] || {});
config.map['*']['BlueAcorn_Module/js/module.bundle'] = 'bundle/BlueAcorn_Module/module.bundle';
/** END: MagentoRequireJsManifestPlugin */"
`;
